sklearn/svm/src/liblinear/linear.cpp

```diff
@@ -70,6 +78,33 @@ static void info(const char *fmt,...)
 static void info(const char *fmt,...) {}
 #endif
 
+// New function to ensure the same behaviour for random number generation on windows and linux
+int myrand() {
+#ifdef _WIN32
+	// In MS Visual Studio (2012) RAND_MAX = 0x7FFF (15bit) = 32767
+	// In Linux GCC (4.6) 32bits RAND_MAX = 0x7FFFFFFF (31bit) = 2147483647
+	// In Linux GCC (4.6) 64bits RAND_MAX = 0x7FFFFFFFFFFFFFFF (63 bits) = 9223372036854775807
+	// so in MS Visual Studio we need to call rand() several times to always ensure the same random number range than in Linux GCC
+	if (std::numeric_limits<int>::max() == 0x7FFFFFFF) {
+		// make a 31bit random number by using several 15bit rand()
+		return ( ( (__int32)rand() << 16) + ( (__int32)rand() << 1) + ( (__int32)rand() >> 14) );
```

Comment:
Am I right in thinking there are parentheses around the entire expression? Why?

Reply:
Are the gains due to randomising the 31st bit worth a third call to rand(), I wonder?

Reply:
I have absoutely no clue.

Note that a much better and faster solution proposed by https://codeforces.com/blog/entry/61587 is to use `mt19937` (32 bits) or `mt19937_64` (64 bits) from `<random>`. This is C++11 though. Do we have the right to use this ?

Reply:
I managed to make it work. Here would be the corresponding code:

```cpp
#include <climits>
#include <random>

// New function to ensure the same behaviour for random number generation on windows and linux
// note that we use always the same seed (12574 here, why not) to get reproducible results
#if INT_MAX == 0x7FFFFFFF
std::mt19937 mt_rand(12574);
#elif INT_MAX == 0x7FFFFFFFFFFFFFFF
std::mt19937_64 mt_rand(12574);
#else
info("Random number generator is not fixed for this system. Please report issue. INT_MAX=%d\n", INT_MAX);
exit(1);
#endif
inline int myrand() {
    // make a 31bit or 63bit positive random number
    return abs( (int)mt_rand());
}
```

Reply:
I commited the fix to see how it goes and surprise: C++11 is not supported in scikit-learn compilation as of today.

`#error This file requires compiler and library support for the ISO C++ 2011 standard. This support must be enabled with the -std=c++11 or -std=gnu++11 compiler options.`

I'll try to see where I can set the flag, just to see how far I can go with this new track. (note that we can always revert to [this version](https://github.com/scikit-learn/scikit-learn/pull/13511/files/1465a56b90abc38ab14bdfed2fa597f2ee72f7a5) if in the end we decide not to go for it).

Reply:
Done. This wakes up a bunch of new test errors. I resolve this conversation and summarize below.